Error while checking build/O2/fullCI for c8bcd4444e25b70e09074685ddee841bef23f986 at 2023-08-24 18:08:
```
## sw/BUILD/O2-latest/log
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
c++: error: unrecognized command-line option '--rtlib=compiler-rt'
/sw/SOURCES/O2/11723-slc8_x86-64/0/Detectors/MUON/MCH/Simulation/src/Response.cxx:125:173: error: expected ';' before ')' token
/sw/SOURCES/O2/11723-slc8_x86-64/0/Detectors/MUON/MCH/Simulation/src/Response.cxx:125:173: error: expected primary-expression before ')' token
ninja: build stopped: subcommand failed.
```
Full log [here](https://ali-ci.cern.ch/alice-build-logs/AliceO2Group/AliceO2/11723/c8bcd4444e25b70e09074685ddee841bef23f986/build_O2_fullCI/pretty.html).

--- a/Detectors/MUON/MCH/Simulation/include/MCHSimulation/Response.h
+++ b/Detectors/MUON/MCH/Simulation/include/MCHSimulation/Response.h
@@ -34,6 +34,7 @@ class Response
   ~Response() = default;
 
   float getChargeSpread() const { return mChargeSpread; }
+  float getPitch() const { return mPitch; }
   float getSigmaIntegration() const { return mSigmaIntegration; }
   bool isAboveThreshold(float charge) const { return charge > mChargeThreshold; }
 
@@ -64,6 +65,21 @@ class Response
   /// compute the number of samples corresponding to the charge in ADC units
   uint32_t nSamples(float charge) const;
 
+  /// Ratio of particle mean eloss with respect MIP's Khalil Boudjemline, sep 2003, PhD.Thesis and Particle Data Book
+  float eLossRatio(float logbetagamma) const;
+  /// ToDo: check Aliroot formula vs PDG, if really log_10 and not ln or bug in Aliroot
+
+  /// Angle effect in tracking chambers at theta =10 degres as a function of ElossRatio (Khalil BOUDJEMLINE sep 2003 Ph.D Thesis) (in micrometers)
+  float angleEffect10(float elossratio) const;
+
+  /// Angle effect: Normalisation form theta=10 degres to theta between 0 and 10 (Khalil BOUDJEMLINE sep 2003 Ph.D Thesis)
+  /// Angle with respect to the wires assuming that chambers are perpendicular to the z axis.
+  float angleEffectNorma(float elossratio) const;
+
+  /// Magnetic field effect: Normalisation form theta=16 degres (eq. 10 degrees B=0) to theta between -20 and 20 (Lamia Benhabib jun 2006 )
+  /// Angle with respect to the wires assuming that chambers are perpendicular to the z axis.
+  float magAngleEffectNorma(float angle, float bfield) const;
+
  private:
   MathiesonOriginal mMathieson{}; ///< Mathieson function
   float mPitch = 0.f;             ///< anode-cathode pitch (cm)
@@ -72,6 +88,8 @@ class Response
   float mSigmaIntegration = 0.f;  ///< number of sigmas used for charge distribution
   float mChargeCorr = 0.f;        ///< amplitude of charge correlation between cathodes
   float mChargeThreshold = 0.f;   ///< minimum fraction of charge considered
+  bool mAngleEffect = true;       ///< switch for angle effect influencing charge deposition
+  bool mMagnetEffect = true;      ///< switch for magnetic field influencing charge deposition
 };
 } // namespace mch
 } // namespace o2

--- a/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/DEDigitizer.cxx
@@ -56,9 +56,25 @@ void DEDigitizer::processHit(const Hit& hit, const InteractionRecord& collisionT
   auto chargeNonBending = charge / chargeCorr;
 
   // local position of the charge distribution
-  math_utils::Point3D<float> pos(hit.GetX(), hit.GetY(), hit.GetZ());
+  auto exitPoint = hit.exitPoint();
+  auto entrancePoint = hit.entrancePoint();
+  math_utils::Point3D<float> lexit{};
+  math_utils::Point3D<float> lentrance{};
+  mTransformation.MasterToLocal(exitPoint, lexit);
+  mTransformation.MasterToLocal(entrancePoint, lentrance);
+
+  auto hitlengthZ = lentrance.Z() - lexit.Z();
+  auto pitch = mResponse.getPitch();
+
   math_utils::Point3D<float> lpos{};
-  mTransformation.MasterToLocal(pos, lpos);
+
+  if (abs(hitlengthZ) > pitch * 1.99) {
+    lpos.SetXYZ((lexit.X() + lentrance.X()) / 2., (lexit.Y() + lentrance.Y()) / 2., (lexit.Z() + lentrance.Z()) / 2.);
+  } else {
+    lpos.SetXYZ(lexit.X(), // take Bragg peak coordinates assuming electron drift parallel to z
+                lexit.Y(), // take Bragg peak coordinates assuming electron drift parallel to z
+                0.0);      // take wire position global coordinate negative
+  }
   auto localX = mResponse.getAnod(lpos.X());
   auto localY = lpos.Y();
 

--- a/Detectors/MUON/MCH/Simulation/src/Response.cxx
+++ b/Detectors/MUON/MCH/Simulation/src/Response.cxx
@@ -89,3 +89,38 @@ uint32_t Response::nSamples(float charge) const
   double signalParam[3] = {14., 13., 1.5};
   return std::round(std::pow(charge / signalParam[1], 1. / signalParam[2]) + signalParam[0]);
 }
+//_____________________________________________________________________
+float Response::eLossRatio(float logbetagamma) const
+{
+  // Ratio of particle mean eloss with respect MIP's Khalil Boudjemline, sep 2003, PhD.Thesis and Particle Data Book
+  /// copied from aliroot AliMUONv1.cxx
+  float eLossRatioParam[5] = {1.02138, -9.54149e-02, +7.83433e-02, -9.98208e-03, +3.83279e-04};
+  return eLossRatioParam[0] + eLossRatioParam[1] * logbetagamma + eLossRatioParam[2] * std::pow(logbetagamma, 2) + eLossRatioParam[3] * std::pow(logbetagamma, 3) + eLossRatioParam[4] * std::pow(logbetagamma, 4);
+}
+//_____________________________________________________________________
+float Response::angleEffect10(float elossratio) const
+{
+  /// Angle effect in tracking chambers at theta =10 degres as a function of ElossRatio (Khalil BOUDJEMLINE sep 2003 Ph.D Thesis) (in micrometers)
+  /// copied from aliroot AliMUONv1.cxx
+  float angleEffectParam[3] = {1.90691e+02, -6.62258e+01, 1.28247e+01};
+  return angleEffectParam[0] + angleEffectParam[1] * elossratio + angleEffectParam[2] * std::pow(elossratio, 2);
+}
+//_____________________________________________________________________
+float Response::angleEffectNorma(float elossratio) const
+{
+  /// Angle effect: Normalisation form theta=10 degres to theta between 0 and 10 (Khalil BOUDJEMLINE sep 2003 Ph.D Thesis)
+  /// Angle with respect to the wires assuming that chambers are perpendicular to the z axis.
+  /// copied from aliroot AliMUONv1.cxx
+  float angleEffectParam[4] = {4.148, -6.809e-01, 5.151e-02, -1.490e-03};
+  return angleEffectParam[0] + angleEffectParam[1] * elossratio + angleEffectParam[2] * std::pow(elossratio, 2) + angleEffectParam[3] * std::pow(elossratio, 3);
+}
+//_____________________________________________________________________
+float Response::magAngleEffectNorma(float angle, float bfield) const
+{
+  /// Magnetic field effect: Normalisation form theta=16 degres (eq. 10 degrees B=0) to theta between -20 and 20 (Lamia Benhabib jun 2006 )
+  /// Angle with respect to the wires assuming that chambers are perpendicular to the z axis.
+  /// copied from aliroot AliMUONv1.cxx
+  float angleEffectParam[7] = {8.6995, 25.4022, 13.8822, 2.4717, 1.1551, -0.0624, 0.0012};
+  float aux = std::abs(angle - angleEffectParam[0] * bfield);
+  return 121.24 / ((angleEffectParam[1] + angleEffectParam[2] * std::abs(bfield)) + angleEffectParam[3] * aux + angleEffectParam[4] * std::pow(aux, 2) + angleEffectParam[5] * std::pow(aux, 3) + angleEffectParam[6] * std::pow(aux, 4));
+}

